lib/Dialect/FIRRTL/FieldRef.cpp

```diff
@@ -0,0 +1,127 @@
+//===- FieldRef.cpp - FIRRTL Field Refs  ------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This header file defines FieldRef and helpers for them.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FieldRef.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace firrtl;
+
+Operation *FieldRef::getDefiningOp() const {
+  if (auto *op = value.getDefiningOp())
+    return op;
+  return value.cast<BlockArgument>().getOwner()->getParentOp();
+}
+
+FieldRef::FieldRef(Value val) : value(val) {
+  // This code walks upwards from the subfield and calculates the field ID at
+  // each level. At each stage, it must take the current id, and re-index it as
+  // a nested bundle under the parent field.. This is accomplished by using the
+  // parent field's ID as a base, and adding the field ID of the child.
+  //
+  // a (0) { b (1) c (2) }
```

Comment:
This was supposed to be an example for field numbering that I never finished. I will delete this.

Reply:
Fixed